Java/core/src/main/java/com/amazon/randomcutforest/genericstore/Store.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.genericstore;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+
+/**
+ * This class provides a fixed amount of storage slots for a given type. When an
+ * object is stored, the class returns an index value that can be used to
+ * retrieve the object later. The purpose of this class is to reduce the overall
+ * memory footprint of an applicaiton by allowing classes to keep index values
+ * instead of pointers. The index values used in this class are ints, which
+ * occupy 4 bytes (Java pointers occupy 8 bytes). For further compression, see
+ * the {@link SmallStore} class which uses 2 byte short integers as index
+ * values.
+ *
+ * @param <T> The type of value being stored.
+ */
+public class Store<T> {
+
+    private final int capacity;
+    private final List<T> store;
+    private final int[] freeBlockStack;
+    private int freeBlockPointer;
+    private final BitSet occupied;
+
+    /**
+     * Create a new store with the given capacity.
+     * 
+     * @param capacity The maximum number of objects that can be added to this
+     *                 store.
+     */
+    public Store(int capacity) {
+        this.capacity = capacity;
+        store = new ArrayList<>(capacity);
+
+        freeBlockStack = new int[capacity];
+        for (int j = 0; j < capacity; j++) {
+            freeBlockStack[j] = capacity - j - 1; // reverse order
+        }
+        freeBlockPointer = capacity - 1;
+        occupied = new BitSet(capacity);
+    }
+
+    /**
+     * @return the maximum number of objects that can be stored.
+     */
+    public int getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * @return the number of objects currently being stored.
+     */
+    public int size() {
+        return capacity - freeBlockPointer - 1;
+    }
+
+    /**
+     * Save a value to this store and return an index which can be used to look up
+     * the value later.
+     * 
+     * @param t The value being stored.
+     * @return an index that can be used to look up the value later.
+     */
+    public int add(T t) {
+        checkState(freeBlockPointer >= 0, "store is full");
+        int index = freeBlockStack[freeBlockPointer--];
+        occupied.set(index);
```

Comment:
question. does it need to check if the index is already set?

Reply:
That's a good question. It _shouldn't_ be possible to reach a state where an index value is available in the freeBlockStack but the occupied bit is set. We could consider adding a check and throwing an illegal state exception for this.

Reply:
since there is the structure for helping maintaining index integrity, it might as well be used. also since the cases this structure protects are also unlikely in correct implementation, i am also ok with discarding it.